### ISSUE TITLE
Which loss is suitable for multitask-regression using skorch？
MSEloss is ok，but if I want to change the weight of each task，what can I do?

### ISSUE THREAD

**Issue by RuanGuojie** [skorch-dev/skorch#856]

**Comment by BenjaminBossan** [comment]
If you have multiple outputs that you want to weigh differently, you could override the `get_loss` method on `NeuralNetRegressor`. However, I think the better approach is to write your own custom loss function. Maybe something like this would work (untested code):

```python
import torch
from torch import nn

class MultiOutputMSELoss(nn.Module):
    def __init__(self, weights):
        super().__init__()
        self.weights=weights

    def forward(self, input, target):
        n = inputs.shape[1]  # number of targets
        assert targets.shape[1] == n
        losses = [nn.functional.mse_loss(input[:, i], target[:, i]) for i in range(n)]
        losses_weighted = [self.weights[i] * losses[i] for i in range(n)]
        return sum(losses_weighted)

# create fake predictions and target: 100 rows, 5 tasks
predictions = torch.rand((100, 5))
targets = torch.rand((100, 5))

# using weights 1 to 5 for the 5 tasks
weights = [1, 2, 3, 4, 5]
criterion = MultiOutputMSELoss(weights)
criterion(predictions, targets)  # tensor(2.4189)
```

You should then be able to pass this as the `criterion` argument to your `NeuralNetRegressor`.

**Comment by RuanGuojie** [comment]
> If you have multiple outputs that you want to weigh differently, you could override the `get_loss` method on `NeuralNetRegressor`. However, I think the better approach is to write your own custom loss function. Maybe something like this would work (untested code):
> 
> ```python
> import torch
> from torch import nn
> 
> class MultiOutputMSELoss(nn.Module):
>     def __init__(self, weights):
>         super().__init__()
>         self.weights=weights
> 
>     def forward(self, input, target):
>         n = inputs.shape[1]  # number of targets
>         assert targets.shape[1] == n
>         losses = [nn.functional.mse_loss(input[:, i], target[:, i]) for i in range(n)]
>         losses_weighted = [self.weights[i] * losses[i] for i in range(n)]
>         return sum(losses_weighted)
> 
> # create fake predictions and target: 100 rows, 5 tasks
> predictions = torch.rand((100, 5))
> targets = torch.rand((100, 5))
> 
> # using weights 1 to 5 for the 5 tasks
> weights = [1, 2, 3, 4, 5]
> criterion = MultiOutputMSELoss(weights)
> criterion(predictions, targets)  # tensor(2.4189)
> ```
> 
> You should then be able to pass this as the `criterion` argument to your `NeuralNetRegressor`.

It works！Thanks a lot！

> weights = [1, 2, 3, 4, 5] 

Does this code represent the relative weight of different tasks？

**Comment by BenjaminBossan** [comment]
> Does this code represent the relative weight of different tasks？

Yes, that's the idea. We take the MSE of each individual task and return the weighted sum. So for these values, the first task (0th column) has half as much weight as the second, a third as much as the third, etc. This is a very simplistic take on the problem, e.g. you may want to normalize the weights to sum up to 1.

**Comment by RuanGuojie** [comment]
> > Does this code represent the relative weight of different tasks？
> 
> Yes, that's the idea. We take the MSE of each individual task and return the weighted sum. So for these values, the first task (0th column) has half as much weight as the second, a third as much as the third, etc. This is a very simplistic take on the problem, e.g. you may want to normalize the weights to sum up to 1.

I see. Thanks again.

**Comment by BenjaminBossan** [comment]
Happy to help. I'll close the issue for now, feel free to re-open if something new comes up with regard to your case.

**Comment by RuanGuojie** [comment]
> weights = [1, 2, 3, 4, 5]
Can I optimize the weight using evolutionary algorithms? Does that make sense? Now I can manually adjust the weights, but I want to find the best weight allocation using scikit-opt/geatpy (multitarget optimization). Could you help me?

**Comment by BenjaminBossan** [comment]
Sorry, I don't believe there is a generally correct answer to this. It really depends on your use case. If you optimize the weights, I would expect that the easiest task will get all the weights and all others get 0, so that's probably not what you want.

> using scikit-opt/geatpy

I don't know what that is, any pointers?

**Comment by RuanGuojie** [comment]
> Sorry, I don't believe there is a generally correct answer to this. It really depends on your use case. If you optimize the weights, I would expect that the easiest task will get all the weights and all others get 0, so that's probably not what you want.
> 
> > using scikit-opt/geatpy
> 
> I don't know what that is, any pointers?

https://github.com/guofei9987/scikit-opt
https://github.com/geatpy-dev/geatpy
These are two libraries of evolutionary algorithms. I want to use it to optimize the weight，but  I am not perfessional.

**Comment by BenjaminBossan** [comment]
I see, thank you for the links.

Again, I think you should have a good understanding of what the use case, of what the final goal is that you want to achieve.

E.g., if you only want to minimize the loss, you could probably use those libraries. You could probably even add the weights to the optimizer so that they are optimized together with the other parameters. But most likely, this is not your real goal, because it will result in the easiest of your tasks getting all the weights.

> but I am not perfessional.

If you're doing this for work or for a university project, I would suggest you discuss what the actual goal is with a more experienced colleague. Maybe together you can figure out what the best approach would be. If there is no further information available, I would just give every task the same weight.

**Comment by RuanGuojie** [comment]
> I see, thank you for the links.
> 
> Again, I think you should have a good understanding of what the use case, of what the final goal is that you want to achieve.
> 
> E.g., if you only want to minimize the loss, you could probably use those libraries. You could probably even add the weights to the optimizer so that they are optimized together with the other parameters. But most likely, this is not your real goal, because it will result in the easiest of your tasks getting all the weights.
> 
> > but I am not perfessional.
> 
> If you're doing this for work or for a university project, I would suggest you discuss what the actual goal is with a more experienced colleague. Maybe together you can figure out what the best approach would be. If there is no further information available, I would just give every task the same weight. 

Yes, I want to minimize the total loss. Therefore, giving every task the same weight maybe not I want, though it is the easiest method. Now I have successfully  added the weights to the optimizer. Thanks for your suggestions.

**Comment by RuanGuojie** [comment]
Yes, I want to minimize the total loss. Therefore, giving every task the same weight maybe not I want, though it is the easiest method. Now I have successfully added the weights to the optimizer. Thanks for your suggestions.

If I only use nn.MSEloss as criterion of  NeuralNetRegressor, the loss is minimized automatically? Does this make it impossible to know the weight of each task?

**Comment by BenjaminBossan** [comment]
To make the weights on the loss learnable, you need to declare them to be `nn.Parameter`s:

```python
class MultiOutputMSELoss(nn.Module):
    def __init__(self, weights):
        super().__init__()
        self.weights = nn.Parameter(torch.tensor(weights, dtype=torch.float32))  # <= changed

    def forward(self, inputs, target):
        n = inputs.shape[1]  # number of targets
        assert targets.shape[1] == n
        losses = [nn.functional.mse_loss(inputs[:, i], target[:, i]) for i in range(n)]
        losses_weighted = [self.weights[i] * losses[i] for i in range(n)]
        return sum(losses_weighted) / sum(self.weights)  # <= changed
```

Then skorch will automatically take care of optimizing these weights as well. Note that I added a normalization term at the end, otherwise the weights would just decline more and more.

But again, depending on your use case, this might not be what you want. You should clarify that this is really solving your problem.

**Comment by RuanGuojie** [comment]
I see. Thanks for your help.
